app/src/main/java/ai/elimu/appstore/synchronization/AppListAdapter.java

```diff
@@ -422,4 +462,61 @@ public ViewHolder(View itemView) {
         }
 
     }
-}
+
+    /**
+     * Get download file url
+     *
+     * @param applicationVersion The application version info of the download file
+     * @return The file url
+     */
+    private String getFileUrl(ApplicationVersion applicationVersion) {
+        Timber.i("applicationVersion.getApplication(): " + applicationVersion.getApplication());
+        Timber.i("applicationVersion.getFileSizeInKb(): " + applicationVersion
+                .getFileSizeInKb());
+        Timber.i("applicationVersion.getFileUrl(): " + applicationVersion.getFileUrl());
+        Timber.i("applicationVersion.getContentType(): " + applicationVersion.getContentType());
+        Timber.i("applicationVersion.getVersionCode(): " + applicationVersion.getVersionCode());
+        Timber.i("applicationVersion.getStartCommand(): " + applicationVersion
+                .getStartCommand());
+        Timber.i("applicationVersion.getTimeUploaded().getTime(): " + applicationVersion
+                .getTimeUploaded().getTime());
+
+        // Download APK file and store it on SD card
+        String fileUrl = BuildConfig.BASE_URL + applicationVersion.getFileUrl() +
+                "?deviceId=" + DeviceInfoHelper.getDeviceId(context) +
+                "&checksum=" + ChecksumHelper.getChecksum(context) +
+                "&locale=" + UserPrefsHelper.getLocale(context) +
+                "&deviceModel=" + DeviceInfoHelper.getDeviceModel(context) +
+                "&osVersion=" + Build.VERSION.SDK_INT +
+                "&applicationId=" + applicationVersion.getApplication().getId() +
+                "&appVersionCode=" + DeviceInfoHelper.getAppVersionCode(context);
+        if (!TextUtils.isEmpty(UserPrefsHelper.getLicenseEmail(context))) {
+            // Custom Project
+            fileUrl += "&licenseEmail=" + UserPrefsHelper.getLicenseEmail(context);
+            fileUrl += "&licenseNumber=" + UserPrefsHelper.getLicenseNumber(context);
+        }
+        Timber.i("fileUrl: " + fileUrl);
+
+        String fileName = applicationVersion.getApplication().getPackageName() + "-" +
+                applicationVersion.getVersionCode() + ".apk";
+        Timber.i("fileName: " + fileName);
+
+        Timber.i("Downloading APK: " + applicationVersion.getApplication().getPackageName() +
+                " (version " + applicationVersion.getVersionCode() + ", " +
+                applicationVersion.getFileSizeInKb() + "kB)");
+
+        return fileUrl;
+    }
+
+    /**
+     * Check if an apk file exists and valid
+     *
+     * @param apkFile      The apk file needs to be checked
+     * @param fileSizeInKb The actually file size of the APK
+     * @return true if apk file exists & valid, false otherwise
+     */
+    private boolean isExistingAndValidApk(@NonNull File apkFile, @NonNull Integer fileSizeInKb) {
+        return apkFile.exists() && (apkFile.length() >= fileSizeInKb * 1024);
```

Comment:
Why is not it enough to only check if the file exists? If an APK file is not valid, it should not be stored in the first place, right?

Reply:
As per my previous comment, Ion library creates an APK file right away when starting a new file download, which makes a present file in the first place. Therefore, we need to check file length as well to make sure the file is fully downloaded.